src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs

```diff
@@ -9,7 +9,7 @@ partial class BoundDagEvaluation
         public override bool Equals(object obj) => obj is BoundDagEvaluation other && this.Equals(other);
         public virtual bool Equals(BoundDagEvaluation other)
         {
```

Comment:
Because we're overriding object.Equals, it's impossible to know statically everywhere this is called.
Instead I've set a breakpoint here, and on GetHashCode, and I've seen where it's been called from in the tests I added.

Namely:
```csharp
DecisionDagBuilder.SimplifyTestsAndBindings(ArrayBuilder<BoundDagTest> testsBuilder, ArrayBuilder<BoundPatternBinding> bindingsBuilder)
     usedValues

DecisionDagBuilder.RemoveEvaluation(RemainingTestsForCase c, BoundDagEvaluation e)
   RemainingTests: c.RemainingTests.WhereAsArray(d => !(d is BoundDagEvaluation e2) || e2 != e),

NullableWalker.LearnFromDecisionDag()
    tempMap

LocalRewriter.PatternLocalRewriter
    _map

```

I was not able to understand what was happening in NullableWalker.LearnFromDecisionDag, so I've simply added tests to verify that my changes do not affect nullability analysis. I imagine @gafter would be able to help here.

In all other cases my changes to Equals seemed correct.

Reply:
Thanks to @jcouv for giving me an overview of how the nullable walker works. It seems to me that this behaviour is correct - since we assume property evaluations are non-mutating, the nullability state of any evaluations of the same property through different types should be the same, so it's ok to use the same slot for them.